### PR TITLE
M4028 Metadata editing UI (backend)

### PR DESCRIPTION
**Pull request by dennishendriksen**
#### Checklist
- [ ] Functionality works & meets specifications
- [ ] Code reviewed
- [ ] Code unit/integration/system tested
- [ ] User documentation updated
- [ ] Test plan template updated
- [ ] Clean commits

### CODE REVIEW

File: molgenis-core-ui/src/main/java/org/molgenis/ui/metadataeditor/MetadataEditorController.java [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+	public EditorEntityTypeResponse createEntityType()
+	{
+		return createEntityTypeResponse();
+	}
+
+	@RequestMapping(value = "/entityType", method = POST)
+	@ResponseStatus(OK)
+	public void upsertEntityType(@RequestBody EditorEntityType editorEntityType)
+	{
+		EntityType entityType = entityTypeMapper.toEntityType(editorEntityType);
```
Comment (Mark-de-Haan) [inline molgenis-core-ui/src/main/java/org/molgenis/ui/metadataeditor/MetadataEditorController.java]: Why are we creating a different EntityType format in the editor compared to the EntityType of MOLGENIS? Using Typed Entities should allow us to keep the format the same all around right?

File: molgenis-core-ui/src/main/java/org/molgenis/ui/metadataeditor/MetadataEditorController.java [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+		return createEntityTypeResponse(editorEntityType);
+	}
+
+	private EditorEntityTypeResponse createEntityTypeResponse(EditorEntityType editorEntityType)
+	{
+		ImmutableList<String> languageCodes = ImmutableList.copyOf(getLanguageCodes().iterator());
+		return EditorEntityTypeResponse.create(editorEntityType, languageCodes);
+	}
+
+	private EditorAttributeResponse createAttributeResponse()
```
Comment (Mark-de-Haan) [inline molgenis-core-ui/src/main/java/org/molgenis/ui/metadataeditor/MetadataEditorController.java]: Similar to the EditorEntityType comment: Why take an Attribute containing all the data you need and changing it into something else?

File: molgenis-core-ui/src/main/java/org/molgenis/ui/metadataeditor/MetadataEditorController.java (hunk 1 above)
Comment (dennishendriksen) [inline molgenis-core-ui/src/main/java/org/molgenis/ui/metadataeditor/MetadataEditorController.java]: The EditorEntityType has a different structure and contains different data such that the frontend can be rendered without having to retrieve parts of the data using multiple REST API requests and piecing them together in the client.

File: molgenis-core-ui/src/main/java/org/molgenis/ui/metadataeditor/MetadataEditorController.java [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+		EntityType entityType = entityTypeMapper.toEntityType(editorEntityType);
+		metadataService.upsertEntityType(entityType);
+	}
+
+	@RequestMapping(value = "/create/attribute", method = GET)
+	@ResponseBody
+	public EditorAttributeResponse createAttribute()
+	{
+		return createAttributeResponse();
+	}
```
Comment (Mark-de-Haan) [inline molgenis-core-ui/src/main/java/org/molgenis/ui/metadataeditor/MetadataEditorController.java]: I am missing an upsertAttribute method, or does the upsertEntityType cover the updating of attributes?

Comment (dennishendriksen) [inline molgenis-core-ui/src/main/java/org/molgenis/ui/metadataeditor/MetadataEditorController.java]: See MetaDataService documentation:
```java
/**
  * Add or update the given entity type and entity type attributes.
  *
  * @param entityType entity type to upsert
  */
 void upsertEntityType(EntityType entityType);
```

File: molgenis-core-ui/src/main/java/org/molgenis/ui/metadataeditor/MetadataEditorController.java [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+	public EditorEntityTypeResponse getEntityType(@PathVariable("id") String entityTypeId)
+	{
+		// metadataService.getEntityType cannot be used due to https://github.com/molgenis/molgenis/issues/5783
+		EntityType entityType = metadataService
+				.getRepository(EntityTypeMetadata.ENTITY_TYPE_META_DATA, EntityType.class).findOneById(entityTypeId);
+
+		return createEntityTypeResponse(entityType);
+	}
+
+	// FIXME currently throws an exception until entity type has an automatically generated identifier
```
Comment (Mark-de-Haan) [inline molgenis-core-ui/src/main/java/org/molgenis/ui/metadataeditor/MetadataEditorController.java]: AUTO_ID for EntityTypes has been merged with the master

**Comment by bartcharbon** [comment]
test this please

**Pull request closed without merging**